akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -111,39 +118,82 @@ private[projection] class SlickProjectionImpl[Offset, Envelope, P <: JdbcProfile
       .via(handlerFlow)
   }
 
+  private val futDone = Future.successful(Done)
+
+  private def applyUserRecovery(envelope: Envelope)(futureCallback: () => Future[Done])(
+      implicit systemProvider: ClassicActorSystemProvider) = {
+
+    implicit val scheduler = systemProvider.classicSystem.scheduler
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
+
+    // this will count as one attempt
+    val firstAttempt = futureCallback()
+
+    firstAttempt.recoverWith {
+      case NonFatal(err) =>
+        eventHandler.onFailure(envelope, err) match {
+          case Fail                         => firstAttempt
+          case Skip                         => futDone
+          case RetryAndFail(retries, delay) =>
+            // less one attempt
+            retry(futureCallback, retries - 1, delay)
+          case RetryAndSkip(retries, delay) =>
+            // less one attempt
+            retry(futureCallback, retries - 1, delay).recoverWith {
+              case NonFatal(_) => futDone
+            }
+        }
+
+    }
+  }
+
   private def processEnvelopeAndStoreOffsetInSameTransaction(env: Envelope)(
-      implicit ec: ExecutionContext): Future[Done] = {
+      implicit systemProvider: ClassicActorSystemProvider): Future[Done] = {
+
     import databaseConfig.profile.api._
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
 
     // run user function and offset storage on the same transaction
     // any side-effect in user function is at-least-once
     val txDBIO =
       offsetStore
         .saveOffset(projectionId, sourceProvider.extractOffset(env))
-        .flatMap(_ => eventHandler(env))
+        .flatMap(_ => eventHandler.handleEvent(env))
```

Comment:
I'm wondering what could be the options here. If we have an exception from 
inside the user function a retry doesn't make sense as it will blow up again. 

Maybe we should have a different treatment. If the strategy is `Skip`, we drop it. That's fine. If not, we should let the whole projection fail. `RetryAndFail` and `RetryAndSkip` doesn't make sense. We will only retry to reach the same bogus handler.

Reply:
I added a `aroundUserHandler` method to take care of this. Not so sure if that's the best solution. 

I'm not finding a case in which it would make sense to run a retry in such a  case. 

I'm holding from writing the tests for it until we decide what are really need here.